lib/web/websocket/stream/websocketstream.js

```diff
@@ -0,0 +1,511 @@
+    this.#wasEverConnected = true
+
+    // 3. Let extensions be the extensions in use .
+    const extensions = parsedExtensions ?? ''
+
+    // 4. Let protocol be the subprotocol in use .
+    const protocol = response.headersList.get('sec-websocket-protocol') ?? ''
+
+    // 5. Let pullAlgorithm be an action that pulls bytes from stream .
+    // 6. Let cancelAlgorithm be an action that cancels stream with reason , given reason .
+    // 7. Let readable be a new ReadableStream .
+    // 8. Set up readable with pullAlgorithm and cancelAlgorithm .
+    const readable = new ReadableStream({
+      start: (controller) => {
+        this.#readableStreamController = controller
+      },
+      pull (controller) {
+        let chunk
+        while (controller.desiredSize > 0 && (chunk = response.socket.read()) !== null) {
+          controller.enqueue(chunk)
+        }
+      },
+      cancel: (reason) => this.#cancel(reason)
+    })
+
+    // 9. Let writeAlgorithm be an action that writes chunk to stream , given chunk .
+    // 10. Let closeAlgorithm be an action that closes stream .
+    // 11. Let abortAlgorithm be an action that aborts stream with reason , given reason .
+    // 12. Let writable be a new WritableStream .
+    // 13. Set up writable with writeAlgorithm , closeAlgorithm , and abortAlgorithm .
+    const writable = new WritableStream({
+      write: (chunk) => this.#write(chunk),
+      close: () => closeWebSocketConnection(this.#handler, null, null),
+      abort: (reason) => this.#closeUsingReason(reason)
+    })
+
+    // Set stream ’s readable stream to readable .
+    this.#readableStream = readable
+
+    // Set stream ’s writable stream to writable .
+    this.#writableStream = writable
+
+    // Resolve stream ’s opened promise with WebSocketOpenInfo «[ " extensions " → extensions , " protocol " → protocol , " readable " → readable , " writable " → writable ]».
+    this.#openedPromise.resolve({
+      extensions,
+      protocol,
+      readable,
+      writable
+    })
+  }
+
+  /** @type {import('../websocket').Handler['onMessage']} */
+  #onMessage (type, data) {
+    // 1. If stream’s ready state is not OPEN (1), then return.
+    if (this.#handler.readyState !== states.OPEN) {
+      return
+    }
+
+    // 2. Let chunk be determined by switching on type:
+    //      - type indicates that the data is Text
+    //          a new DOMString containing data
+    //      - type indicates that the data is Binary
+    //          a new Uint8Array object, created in the relevant Realm of the
+    //          WebSocketStream object, whose contents are data
+    let chunk
+
+    if (type === opcodes.TEXT) {
+      try {
+        chunk = utf8Decode(data)
+      } catch {
+        failWebsocketConnection(this.#handler, 'Received invalid UTF-8 in text frame.')
+        return
+      }
+    } else if (type === opcodes.BINARY) {
+      chunk = new Uint8Array(data.buffer, data.byteOffset, data.byteLength)
```

Comment:
```suggestion
      chunk = data.byteLength === data.buffer.byteLength ? new Uint8Array(data.buffer, data.byteOffset, data.byteLength) : new Uint8Array(data)
```

Copy buffer if needed.